File: src/test/java/com/jillesvangurp/geo/GeoHashUtilsTest.java
/**
 * Copyright (c) 2012, Jilles van Gurp
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.jillesvangurp.geo;

import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import java.util.Set;

import static com.jillesvangurp.geo.GeoGeometry.*;
import static com.jillesvangurp.geo.GeoHashUtils.*;
import static java.lang.Math.abs;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.greaterThan;
import static org.hamcrest.Matchers.lessThan;

@Test
public class GeoHashUtilsTest {

    @DataProvider
    public Object[][] coordinates() {
        return new Object[][]{
                {0.1, -0.1, "ebpbtdpntc6e"}, // very cold there ;-)
                {52.530888, 13.394904, "u33dbfcyegk2"}, // home sweet home
        };
    }

    @Test(dataProvider = "coordinates")
    public void shouldDecode(Double lat, Double lon, String geoHash) {
        double[] decoded = decode(geoHash);
        assertSimilar(lat, decoded[1]);
        assertSimilar(lon, decoded[0]);
    }

    @Test(dataProvider = "coordinates")
    public void shouldEncode(Double lat, Double lon, String expectedGeoHash) {
        assertThat(encode(lat, lon, DEFAULT_GEO_HASH_LENGTH), is(expectedGeoHash));
        assertThat(encode(new double[]{lon, lat}), is(expectedGeoHash));
    }

    @Test(dataProvider = "coordinates")
    public void shouldContainCoordinate(Double lat, Double lon, String geoHash) {
        assertThat("hash should contain the coordinate",
                contains(geoHash, lat, lon));
        assertThat("hash should not contain the swapped coordinate",
                !contains(geoHash, lon, lat));
    }

    @Test(dataProvider = "coordinates")
    public void shouldDecodeBbox(Double lat, Double lon, String geoHash) {
        double[] bbox = decodeBbox(geoHash);
        assertThat(abs((bbox[0] + bbox[2]) / 2 - lon), lessThan(0.0001));
        assertThat(abs((bbox[1] + bbox[3]) / 2 - lat), lessThan(0.0001));
    }

    @Test(dataProvider = "coordinates")
    public void shouldCalculateEast(Double lat, Double lon, String geoHash) {
        String original = geoHash.substring(0, 3);
        String calculated = GeoHashUtils.east(geoHash.substring(0, 3));
        assertThat("east hash should not contain the coordinate",
                !contains(calculated, lat, lon));
        double[] bbox = decodeBbox(original);
        double[] eastBbox = decodeBbox(calculated);
        assertSimilar(bbox[1], eastBbox[1]);
        assertSimilar(bbox[3], eastBbox[3]);
        assertSimilar(bbox[2], eastBbox[0]);
        assertSimilar((eastBbox[2] - bbox[0]) / 2, bbox[2] - bbox[0]);
        double nl = GeoHashUtils.decode(calculated)[0];
        double ol = GeoHashUtils.decode(original)[0];
        assertThat("decoded hash lon should be east of original", GeoHashUtils.isEast(nl, ol));
    }

    @Test(dataProvider = "coordinates")
    public void shouldCalculateSouth(Double lat, Double lon, String geoHash) {
        String original = geoHash.substring(0, 3);
        String calculatedHash = GeoHashUtils.south(original);
//        System.out.println(original + " " + calculatedHash);
        double[] oBox = decodeBbox(original);
        double[] cBox = decodeBbox(calculatedHash);
        assertThat("calculated hash should not contain the coordinate",
                !contains(calculatedHash, lat, lon));
        double oWest = oBox[0];
        double oSouth = oBox[1];
        double oEast = oBox[2];
        double oNorth = oBox[3];
        double cWest = cBox[0];
        double cSouth = cBox[1];
        double cEast = cBox[2];
        double cNorth = cBox[3];

        assertSimilar((oNorth - cSouth) / 2, oNorth - oSouth);
        assertSimilar(oSouth, cNorth);
        assertSimilar(oEast, cEast);
        assertSimilar(oWest, cWest);
        double nl = GeoHashUtils.decode(calculatedHash)[1];
        double ol = GeoHashUtils.decode(original)[1];
        assertThat("decoded hash lat should be south of original", GeoHashUtils.isSouth(nl, ol));
    }

    @Test(dataProvider = "coordinates")
    public void shouldCalculateNorth(Double lat, Double lon, String geoHash) {
        String original = geoHash.substring(0, 3);
        String calculatedHash = GeoHashUtils.north(original);
        double[] oBox = decodeBbox(original);
        double[] cBox = decodeBbox(calculatedHash);
        assertThat("calculated hash should not contain the coordinate",
                !contains(calculatedHash, lat, lon));
        double oWest = oBox[0];
        double oSouth = oBox[1];
        double oEast = oBox[2];
        double oNorth = oBox[3];
        double cWest = cBox[0];
        double cSouth = cBox[1];
        double cEast = cBox[2];
        double cNorth = cBox[3];
        assertSimilar(oNorth, cSouth);
        assertSimilar((oSouth - cNorth) / 2, oSouth - oNorth);
        assertSimilar(oWest, cWest);
        assertSimilar(oEast, cEast);
//        System.out.println();
        double nl = GeoHashUtils.decode(calculatedHash)[1];
        double ol = GeoHashUtils.decode(original)[1];
        assertThat("decoded hash lat should be north of original", GeoHashUtils.isNorth(nl, ol));
    }

    @Test(dataProvider = "coordinates")
    public void shouldCalculateWest(Double lat, Double lon, String geoHash) {
        String original = geoHash.substring(0, 3);
        String calculatedHash = GeoHashUtils.west(original);
        double[] oBox = decodeBbox(original);
        double[] cBox = decodeBbox(calculatedHash);
        assertThat("calculated hash should not contain the coordinate",
                !contains(calculatedHash, lat, lon));
        double oWest = oBox[0];
        double oSouth = oBox[1];
        double oEast = oBox[2];
        double oNorth = oBox[3];
        double cWest = cBox[0];
        double cSouth = cBox[1];
        double cEast = cBox[2];
        double cNorth = cBox[3];

        assertSimilar(oSouth, cSouth);
        assertSimilar(oNorth, cNorth);
        assertSimilar((oEast - cWest) / 2, oEast - oWest);
        assertSimilar(oWest, cEast);
        double nl = GeoHashUtils.decode(calculatedHash)[0];
        double ol = GeoHashUtils.decode(original)[0];
        assertThat("decoded hash lon should be west of original", GeoHashUtils.isWest(nl, ol));
    }

    private void assertSimilar(double d1, double d2) {
        // allow for some margin of error
        assertThat("should be similar" + d1 + " and " + d2, abs(d1 - d2),
                lessThan(0.0000001));
    }

//    @Test(enabled = false)
//    public void shouldCalculateBboxSizes() {
//        System.out.println("<table border=\"1\">");
//        System.out.println("<th><td>latitude</td><td>1</td><td>2</td><td>3</td><td>4</td><td>5</td><td>6</td><td>7</td><td>8</td><td>9</td><td>10</td><td>11</td><td>12</td></th>");
//        printHashSizes(90, 0);
//        printHashSizes(80, 0);
//        printHashSizes(70, 0);
//        printHashSizes(60, 0);
//        printHashSizes(50, 0);
//        printHashSizes(40, 0);
//        printHashSizes(30, 0);
//        printHashSizes(20, 0);
//        printHashSizes(10, 0);
//        printHashSizes(0, 0);
//        System.out.println("</table>");
//    }

//    private void printHashSizes(double lat, double lon) {
//        String geoHash = encode(lat, lon, DEFAULT_GEO_HASH_LENGTH);
//
//        // not a test but nice to get a sense of the scale of a geo hash
//        System.out.println("<tr><td>" + lat + "</td>");
//        for (int i = 1; i <= geoHash.length(); i++) {
//            String prefix = geoHash.substring(0, i);
//            double[] bbox = decodeBbox(prefix);
//            double vertical = roundToDecimals(distance(bbox[0], bbox[3], bbox[1], bbox[3]), 2);
//            double horizontal = roundToDecimals(distance(bbox[0], bbox[2], bbox[0], bbox[3]), 2);
//            System.out.print("<td>" + horizontal + "x" + vertical + "</td>");
//        }
//        System.out.print("</tr>\n");
//    }

    @Test(enabled = false)
    public void shouldCalculateSubHashesForHash() {
        String hash = "u33dbfc";
        String[] subHashes = GeoHashUtils.subHashes(hash);
        assertThat(subHashes.length, is(32));
        String first = subHashes[0];
        String row = first;
        for (int j = 0; j < 16; j++) {
            String column = row;
            for (int i = 0; i < 8; i++) {
//                System.out.print(column + " ");
                column = east(column);
            }
//            System.out.println();
            row = north(row);
        }
    }

    @DataProvider
    public Object[][] lines() {
        return new Object[][]{
                {1, 1, 2, 2, "/"},
                {2, 2, 1, 1, "\\"},
                {2, 1, 1, 1, "|"},
                {1, 2, 1, 1, "-"},
                {1, 1, 2, 1, "|"},
                {1, 1, 1, 2, "-"},
                {1, 1, 1, 2, "|"}};
    }

    @Test(dataProvider = "lines")
    public void shouldCalculateHashesForLine(double lat1, double lon1,
                                             double lat2, double lon2, String orientation) {
        Set<String> hashes = GeoHashUtils.geoHashesForLine(10000, lat1, lon1, lat2,
                lon2);

//        GsonBuilder b = new GsonBuilder();
//        Gson gson = b.serializeNulls().create();
//
//
//        PointGeometry p1 = PointGeometry.of(lon1, lat1);
//        PointGeometry p2 = PointGeometry.of(lon2, lat2);
//        double[][] line = p1.line(p2);
//        LineStringGeometry lineGeo = new LineStringGeometry(line,null);
//        System.out.println(gson.toJson(
//                FeatureCollection.of(p1.asFeature(null,null), p2.asFeature(null,null), lineGeo.asFeature(null,null)).plus( FeatureCollection.fromGeoHashes(hashes))));

        assertThat("number of hashes, orientation " + orientation, hashes.size(), greaterThan(10));

    }


    @DataProvider
    public Object[][] samplePoints() {
        return new Double[][]{
                {10.0, 85.0, 15.0},
                {10.0, 50.0, 15.0},
                {10.0, 0.0, 15.0},
                {100.0, 85.0, 15.0},
                {100.0, 50.0, 15.0},
                {100.0, 0.0, 15.0},
                {1000.0, 85.0, 15.0},
                {1000.0, 50.0, 15.0},
                {1000.0, 0.0, 15.0},
                {10000.0, 85.0, 15.0},
                {10000.0, 50.0, 15.0},
                {10000.0, 0.0, 15.0},
                {100000.0, 85.0, 15.0},
                {100000.0, 50.0, 15.0},
                {100000.0, 0.0, 15.0}
        };
    }

    @Test(dataProvider = "samplePoints")
    public void shouldCalculateHashLength(double m, double latitude, double longitude) {
        int length = GeoHashUtils.suitableHashLength(m, latitude, longitude);
        String hash = encode(latitude, longitude, length);
        double[] bbox = decodeBbox(hash);
        double distance = GeoGeometry.distance(bbox[0], bbox[1], bbox[0], bbox[3]);
        assertThat(distance, lessThan(m));
    }
}

File: src/test/kotlin/com/jillesvangurp/geogeometry/GeoHashUtilsTest.kt
package com.jillesvangurp.geogeometry

import com.jillesvangurp.geo.GeoGeometry.Companion.distance
import com.jillesvangurp.geo.GeoGeometry.Companion.overlap
import com.jillesvangurp.geo.GeoHashUtils
import com.jillesvangurp.geo.GeoHashUtils.Companion.decode
import com.jillesvangurp.geo.GeoHashUtils.Companion.decodeBbox
import com.jillesvangurp.geo.GeoHashUtils.Companion.east
import com.jillesvangurp.geo.GeoHashUtils.Companion.encode
import com.jillesvangurp.geo.GeoHashUtils.Companion.geoHashesForCircle
import com.jillesvangurp.geo.GeoHashUtils.Companion.geoHashesForPolygon
import com.jillesvangurp.geo.GeoHashUtils.Companion.isEast
import com.jillesvangurp.geo.GeoHashUtils.Companion.isWest
import com.jillesvangurp.geo.GeoHashUtils.Companion.west
import com.jillesvangurp.geo.eastLongitude
import com.jillesvangurp.geo.latitude
import com.jillesvangurp.geo.longitude
import com.jillesvangurp.geo.northLatitude
import com.jillesvangurp.geo.southLatitude
import com.jillesvangurp.geo.westLongitude
import io.kotest.data.blocking.forAll
import io.kotest.data.row
import io.kotest.matchers.shouldBe
import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.jupiter.api.Test
import org.testng.Assert
import kotlin.math.abs

class GeoHashUtilsTest {
    val coordinatesWithHashes = arrayOf(
        row(0.1, -0.1, "ebpbtdpntc6e"),
        row(52.530888, 13.394904, "u33dbfcyegk2")
    )

    fun lines() = arrayOf(
        arrayOf(1, 1, 2, 2),
        arrayOf(2, 2, 1, 1),
        arrayOf(2, 1, 1, 1),
        arrayOf(1, 2, 1, 1),
        arrayOf(1, 1, 2, 1),
        arrayOf(1, 1, 1, 2),
        arrayOf(1, 1, 1, 2)
    )



    @Test
    fun `decode hash`() {
        forAll(*coordinatesWithHashes) { lat: Double, lon: Double, geoHash: String ->
            val decoded = GeoHashUtils.decode(geoHash)
            decoded.latitude shouldBeApproximately lat
            decoded.longitude shouldBeApproximately lon
        }
    }

    @Test
    fun `encode hash`() {
        forAll(*coordinatesWithHashes) { lat: Double, lon: Double, geoHash: String ->
            GeoHashUtils.encode(lat, lon) shouldBe geoHash
            GeoHashUtils.encode(doubleArrayOf(lon, lat)) shouldBe geoHash
        }
    }

    @Test
    fun `hash bbox should contain coordinate`() {
        forAll(*coordinatesWithHashes) { lat: Double, lon: Double, geoHash: String ->
            GeoHashUtils.contains(geoHash, lat, lon) shouldBe true
            GeoHashUtils.contains(geoHash, lon, lat) shouldBe false
        }
    }

    @Test
    fun `decode bbox`() {
        forAll(*coordinatesWithHashes) { lat: Double, lon: Double, geoHash: String ->
            val bbox = GeoHashUtils.decodeBbox(geoHash)

            abs((bbox.southLatitude + bbox.northLatitude) / 2 - lat) shouldBeApproximately 0.0
            abs((bbox.westLongitude + bbox.eastLongitude) / 2 - lon) shouldBeApproximately 0.0
        }
    }

    @Test
    fun shouldCalculateEastOn180() {
        val hash = encode(-18.0, 179.9, 3)
        var bbox = decodeBbox(hash)
        MatcherAssert.assertThat(
            bbox[2],
            CoreMatchers.equalTo(180.0)
        )
        val east = east(hash)
        bbox = decodeBbox(east)
        MatcherAssert.assertThat(
            bbox[0],
            CoreMatchers.equalTo(-180.0)
        )
    }

    @Test
    fun shouldCalculateWestOn180() {
        val hash = encode(-18.0, -179.9, 3)
        var bbox = decodeBbox(hash)
        MatcherAssert.assertThat(
            bbox[0],
            CoreMatchers.equalTo(-180.0)
        )
        val west = west(hash)
        bbox = decodeBbox(west)
        MatcherAssert.assertThat(
            bbox[2],
            CoreMatchers.equalTo(180.0)
        )
    }


    @Test
    fun shouldCalculateHashesForPolygon() {
        val polygon = arrayOf(
            doubleArrayOf(-1.0, 1.0),
            doubleArrayOf(2.0, 2.0),
            doubleArrayOf(3.0, -1.0),
            doubleArrayOf(-2.0, -4.0)
        )
        var min = 10
        val geoHashesForPolygon = geoHashesForPolygon(
            8, *polygon
        )
        for (h in geoHashesForPolygon) {
            min = Math.min(min, h.length)
        }
        MatcherAssert.assertThat(
            "there should be some hashes with length=3",
            min,
            CoreMatchers.`is`(3)
        )
        MatcherAssert.assertThat(
            "huge area, should generate lots of hashes",
            geoHashesForPolygon.size > 1000
        )
    }

    @Test
    fun shouldCalculateHashesForCircle() {
        val hashesForCircle =
            geoHashesForCircle(8, 52.0, 13.0, 2000.0)
        for (hash in hashesForCircle) {
            val point = decode(hash)
            val distance = distance(point, doubleArrayOf(13.0, 52.0))
            MatcherAssert.assertThat(
                distance,
                Matchers.lessThan(2000.0)
            )
        }
    }

    @Test
    fun shouldCheckIfWest() {
        MatcherAssert.assertThat("should be west", isWest(90.0, 91.0))
        MatcherAssert.assertThat("should be west", isWest(-1.0, 1.0))
        MatcherAssert.assertThat("should be west", isWest(-89.0, 90.0))
        MatcherAssert.assertThat("should be west", isWest(180.0, -178.0))
        MatcherAssert.assertThat("should be west", isWest(180.0, -179.99527198651967))
        MatcherAssert.assertThat("should not be west", !isWest(-179.0, 180.0))
        MatcherAssert.assertThat("should not be west", !isWest(91.0, 90.0))
        MatcherAssert.assertThat("should not be west", !isWest(-179.0, 180.0))
        MatcherAssert.assertThat("should not be west", !isWest(89.0, -90.0))
        MatcherAssert.assertThat("should not be west", !isWest(1.0, -1.0))
        MatcherAssert.assertThat("should not be west", !isWest(91.0, 90.0))
        MatcherAssert.assertThat("should not be west", !isWest(-91.0, 90.0))
    }

    @Test
    fun shouldCheckIfEast() {
        MatcherAssert.assertThat("should not be east", !isEast(90.0, 91.0))
        MatcherAssert.assertThat("should not be east", !isEast(-1.0, 1.0))
        MatcherAssert.assertThat("should not be east", !isEast(-89.0, 90.0))
        MatcherAssert.assertThat("should not be east", !isEast(180.0, -178.0))
        MatcherAssert.assertThat("should not be west", !isEast(180.0, -179.99527198651967))
        MatcherAssert.assertThat("should be east", isEast(-179.0, 180.0))
        MatcherAssert.assertThat("should be east", isEast(91.0, 90.0))
        MatcherAssert.assertThat("should be east", isEast(-179.0, 180.0))
        MatcherAssert.assertThat("should be east", isEast(89.0, -90.0))
        MatcherAssert.assertThat("should be east", isEast(1.0, -1.0))
        MatcherAssert.assertThat("should be east", isEast(91.0, 90.0))
        MatcherAssert.assertThat("should be east", isEast(-91.0, 90.0))
    }

    @Test
    fun shouldBeNeitherWestNorEast() {
        MatcherAssert.assertThat("should not be west", !isWest(-90.0, 90.0))
        MatcherAssert.assertThat("should not be east", !isEast(-90.0, 90.0))
    }

    @Test
    fun shouldGenerateCircleHashesThatAreAllWithinRadiusOfCircle() {
        val latitude = 52.529731
        val longitude = 13.401284
        val radius = 500
        val hashes =
            geoHashesForCircle(8, latitude, longitude, radius.toDouble())
        for (hash in hashes) {
            MatcherAssert.assertThat(
                distance(
                    decode(
                        hash
                    ), doubleArrayOf(longitude, latitude)
                ), Matchers.lessThan(500.0)
            )
        }
        MatcherAssert.assertThat(hashes.size, Matchers.greaterThan(radius))
    }

    @Test
    fun shouldOverlap() {
        val polygon = arrayOf(
            doubleArrayOf(50.0, 15.0),
            doubleArrayOf(53.0, 15.0),
            doubleArrayOf(53.0, 11.0),
            doubleArrayOf(50.0, 11.0)
        )
        val p2overlapping = arrayOf(
            doubleArrayOf(51.0, 16.0),
            doubleArrayOf(52.0, 16.0),
            doubleArrayOf(52.0, 10.0),
            doubleArrayOf(51.0, 10.0)
        )
        val p3outside = arrayOf(
            doubleArrayOf(60.0, 15.0),
            doubleArrayOf(63.0, 15.0),
            doubleArrayOf(63.0, 11.0),
            doubleArrayOf(60.0, 11.0)
        )
        val p4inside = arrayOf(
            doubleArrayOf(51.0, 14.0),
            doubleArrayOf(52.0, 14.0),
            doubleArrayOf(52.0, 12.0),
            doubleArrayOf(51.0, 12.0)
        )
        Assert.assertTrue(overlap(polygon, polygon))
        Assert.assertTrue(overlap(polygon, p2overlapping))
        Assert.assertTrue(overlap(p2overlapping, polygon))
        Assert.assertFalse(overlap(p3outside, polygon))
        Assert.assertTrue(overlap(polygon, p4inside))
        Assert.assertTrue(overlap(p4inside, polygon))
    }
}
